haskell/nix/default.nix

```diff
@@ -0,0 +1,113 @@
+/**
+  Generate a bazel-friendly nix package containing
+  - The haskell package itself
+  - Its documentation
+  - A bazel file ready to be loaded from the `BUILD` file and containing the
+  right call to `haskell_import`
+*/
+{ runCommand, lib, writeTextDir, symlinkJoin }:
+let
+  /* Generate the BUILD file for the package */
+  genBuildFile =
+    { package_name, package, ghc }:
+    runCommand "${package.name}-BUILD" {
+      ghc_pkg = "${ghc}/bin/ghc-pkg --simple-output -v0";
+      GHC_PACKAGE_PATH = "${package}/lib/${ghc.name}/package.conf.d";
+      inherit package_name;
+    } ''
+      query_field () {
+        $ghc_pkg field ${package_name} $@
+      }
+
+      query_haddock () {
+        echo -n "["
+        for FIELD in $(query_field $@); do
+          echo -n '"'
+          echo -n $(echo "$FIELD" | cut -d/ -f5-)
+          echo -n '*",'
+        done
+        echo -n "]"
+      }
+
+      query_list () {
+        echo -n "["
+        for FIELD in $(query_field $@); do
+          echo -n '"'
+          echo -n $(echo "$FIELD" | cut -d/ -f5-)
+          echo -n '",'
+        done
+        echo -n "]"
+      }
+
+      get_deps () {
+        echo -n "["
+        for DEP in $(query_field depends); do
+          DEPNAME=$(echo $DEP | sed 's/-[0-9].*//')
+          if [[ -n $DEPNAME ]]; then
+            echo -n "\"@hackage-$DEPNAME\","
+          fi
+        done
+        echo -n "]"
+      }
+
+      mkdir -p $out
+      cat <<EOF > $out/BUILD.bzl
+      load("@io_tweag_rules_haskell//haskell:import.bzl", haskell_import_new = "haskell_import")
+      deps_ids = $(query_list depends)
+      deps_repos = $(get_deps)
```

Comment:
So this has the potential to cause a dependency cycle it seems. When building `haddock-library` (e.g. with `bazel build @hackage-haddock-library//...`), this list includes `haddock-library` itself. I've not worked out if this is a quirk of the documentation generation process or the Nix derivation for `haddock-library`, but we may need a workaround here.

Reply:
There is something strange going on there indeed. It looks like the `haddock-library` package has several libraries, and that apparently perturbs the script.
I can't get the nix package to build though, which prevents me from testing it further, which nixpkgs revision are you using (or maybe you have a custom override for it?)

Reply:
The issue seems to come from [here](https://github.com/tweag/rules_haskell/pull/442/files/8fadda46c00a4d6fbcf7cca7e9d96b073b666ccf#diff-0415aa1dd12e257be8d4ff76b830094fR45): I've assumed that package-ids were always in the form `{package_name}-{version}-{some_random_garbage}`, but this isn't true when package define "internal libraries" whose ids are of the form `{package_name}-{version}-{some_random_garbage}-{library_name}` (in this example, `haddock-library-1.5.0.1-919WkNWGnku6Kvc4vlhbv1` depends on `haddock-library-1.5.0.1-919WkNWGnku6Kvc4vlhbv1-attoparsec`, but the script doesn't make any difference between both).

Since the internal libraries are in the same package as their main library, there is no need to distinguish them so the TL;DR; is that we can just filter-out self-dependencies and things should work smoothly

Reply:
My last commit should fix this, though I couldn't test it on `haddock-library` as I can't build it

Reply:
Nice -- I noticed this when inspecting the `ghc-pkg` DB for that derivation but didn't clock the subpackage. The latest commit resolves this for me :smile: